feat(sql): disk spill support for FastMap

hilmialf opened the pull request:
WIP for https://github.com/questdb/questdb/issues/3848
Not ready for review, but created PR early for feedback

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=3877) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=3877) before we can accept your contribution.<br/><hr/>**hilmi alfatih** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=3877) it.</sub>

hilmialf commented:
@puzpuzpuz Hi Andrei, I am trying to switch implementation to use `MemoryCARWImpl` first before adding mechanism to switch to mmap version.

It currently still has memory leak since I do not check for `maxResize` and `MAX_HEAP_SIZE` limit yet.
If the current PR state is in the good direction, I'll proceed as follows:
1. Add additional private vars inside MemoryCARWImpl (nResize, MAX_HEAP_SIZE) and perform check before each resize to obey the new limitation.
2. Continue the work on implementing the mmap part.

looking forward to your feedback.

puzpuzpuz commented:
> 1. Add additional private vars inside MemoryCARWImpl (nResize, MAX_HEAP_SIZE) and perform check before each resize to obey the new limitation.

I'd avoid modifying `MemoryCARWImpl` - this class already has sufficient general-purpose API. The checks can be made externally, i.e. in `FastMap`.

> but MemoryCARWImpl does not seem to extend the same way as original FastMap impl (<<1 for every extend).

That's also fine - as a user, I wouldn't be happy if the `FastMap` file grows from 4GB to 8GB in one go while it really needs 5GB. `FastMap` should employ different growth strategy based on the heap type (file-baked or not).

hilmialf commented:
@puzpuzpuz could you take a quick look on current PR state?
Here is what's changed:
- `pageSize` is renamed to `heapSize` inside FastMap to make the name clearer.
- created `MemoryCARWSpillable` along with unit tests. *let me revisit for the thoroughness, right now I just copied directly from `MemoryCMARImplTest`.

I would like your suggestion on the followings:
- what should be the best default `pageSize` for `CARWImpl` inside `MemoryCARWSpillable`
- similarly, what should be the best default `pageSize` for `CMARWImpl` inside `MemoryCARWSpillable`
- Also, I wonder whether we need to make enforcement if both `pageSize` should be equal or not

In addition, I found several parts where (to me, CMIIW) looks inconsistent. I marked them with `// TODO`.
for instance:
- extendedSegmentSize definition.

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+        if (!isSpilled && (sz > diskSpillThreshold)){
+
+            MemoryCARW cmarw = initCMARW(sz, mem.getAppendOffset());
+            synchronized (this){
```
Why is this `synchronized` required? `Memory*` and `FastMap` classes are not thread-safe and supposed to be used on a single thread.

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+            MemoryCARW cmarw = initCMARW(sz, mem.getAppendOffset());
+            synchronized (this){
+                Vect.memmove(cmarw.getAddress(), mem.getAddress(), mem.size());
+                LOG.debug().$("Start spilling. curSz: ").$(mem.size()).$(", requestedSz: ").$(sz).$(", newSz: ").$(cmarw.size());
```
A trailing `.$()`/`.I$()` call must end any logging call:
```suggestion
                LOG.debug().$("Start spilling. curSz: ").$(mem.size()).$(", requestedSz: ").$(sz).$(", newSz: ").$(cmarw.size()).$();
```

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+        return mem.getAddress() + mem.getAppendOffset();
+    }
+
+    private MemoryCARWImpl initCARW() {
```
nit: the `initCARW`/`initCMARW` names are too cryptic. Better call them `initFileBasedMem()`/`initMem()` or something like that.

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/api/MemoryCARW.java:
```diff
@@ -227,12 +253,27 @@ default void putStr(long offset, CharSequence value) {
         }
     }
 
+    default void putStrLowerCase(long offset, CharSequence value) {
```
Considering that only FastMap needs this method, it might be a better idea to implement it in `MemoryCARWSpillable`. Same with `putBin32` and other methods.

puzpuzpuz commented on core/src/test/java/io/questdb/test/cairo/vm/MemoryCARWSpillableTest.java:
```diff
@@ -0,0 +1,1247 @@
+
+import static org.junit.Assert.*;
+
+public class MemoryCARWSpillableTest extends AbstractCairoTest {
```
There should be some tests that assert the disk spill case (the file gets created when the memory grows and the memory contents are retained, the file gets deleted when the memory shrinks and so on). Also, some negative case tests would be great, e.g. if file creation fails, a meaningful exception is thrown; if file deletion fails on map shrink, an exception is thrown (yet, `FastMap` may ignore it in this scenario as we need to delete the temp dir on startup).

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+import io.questdb.std.Vect;
+import io.questdb.std.str.Path;
+
+public class MemoryCARWSpillable implements MemoryCARW, Mutable {
```
This class is `FastMap` specific, so I'd move it to the `io.questdb.cairo.map` package and give it a more obvious name like `SpillableMapHeap` or `SpillableHeapMemory`.

hilmialf replied:
I see. I was thinking if it can be used more generally. But anyway, let me move if no other classes would need it.

puzpuzpuz replied:
Introducing an abstraction preliminary, i.e. without a few use cases around it, usually doesn't end up with something maintainable. That's why it's better not to introduce another general-purpose memory class until we have a few use cases for it. ATM we only have one use case, so it's better to write something suitable for that use case.

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+    private String tmpFile;
+
+    // TODO: For now assume the same pageSize
+    public MemoryCARWSpillable(long pageSize, int maxPages, long diskSpillThreshold, CairoConfiguration cairoConf){
```
No need to have so many constructors. It's enough to keep only the one needed by the `FastMap`. The same applies to all methods here - they can be `FastMap`-specific and there is no need to keep them if they're unused. In fact, this call doesn't even have to implement `MemoryCARW` especially considering that certain methods lose their meaning due to different underlying memories, e.g. `getExtendSegmentSize()`, `getPageCount()` or `getPageSize()`.

hilmialf replied:
I see..
Tbh, I found several inconsistencies inside MemoryCMARWImpl and MemoryCARWImpl where it might refer to different things (but most of the methods there such as `getExtendSegmentSize()` are derived from shared interface, so I believe they should convey the same meaning)

hilmialf replied:
Btw, how should I set the pageSize for each memory type (filebased and plain mem)?
Is it configurable?
if so, what should be sensible default? 4kb?

puzpuzpuz replied:
For the "normal" (non-file-baked) memory, the page size should be the same as `FastMap`'s initial size. For the mmapped memory, page size doesn't matter - the extend segment size does. I'd make both of these configurable, so that `FastMap` passes them when initializing the heap object. Also, the extend segment size should be configurable through the `CairoConfiguration`. The default value may be something like 16MB since we know that disk spill has to deal with large files and frequent file extensions will be costly with 4KB.

puzpuzpuz commented on core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java:
```diff
@@ -0,0 +1,283 @@
+        this.spillMemoryTag = spillMemoryTag;
+        this.mem = initCARW();
+        this.isSpilled = false;
+        this.tmpFile = UUID.randomUUID() + ".tmp";
```
The file name and path should be provided externally, i.e. from the `FastMap`. This class shouldn't bother with file name generation.

puzpuzpuz reviewed (commented):
The overall direction looks good. Left a few comments here and there.

hilmialf commented:
Hi Andrei @puzpuzpuz. My apology for late follow-up 🙇 

I updated in several places. Here is highlight from the last time:
1. Added necessary Configuration
2. Updated places where FastMap is instantiated to use the new constructor.

Left to do:
1. Add test cases for spillable FastMap. Notes:
- I did extensive testcase for SpillableHeapMemory in SpillableHeapMemoryTest. However, I have not done test yet from FastMapTest. Instead, I keep the current FastMapTest test case as it is with old FastMap constructor (unspillable memory).

I would like to hear your opinion, especially about the current config and settings.
In addition, I found the behavior of extend() for `MemoryCARWImpl` and `MemoryCMARWImpl` are different internally when the new size is multiple of pageSize. In current implementation, I somehow modify them so that they behave similarly. You can find the detail on the `SpillableHeapMemory` `checkAndExtend()` part. Please let me know if this is not what you thought it should be.

Thanks a lot for your time.

P.S.
Oops my formatter is unbehaving.. its somehow restructuring in unnecessary part of the code. Could you point out where I setup wrong with the formatter? I tried to follow the development guide.

puzpuzpuz commented:
> Oops my formatter is unbehaving.. its somehow restructuring in unnecessary part of the code. Could you point out where I setup wrong with the formatter? I tried to follow the development guide.

If you're using IntelliJ IDEA, it should pick up the code style automatically and it's checked in the repo. Personally I use the following actions on save, so that the edited files are always auto-formatted.
![Screenshot from 2023-11-13 11-49-14](https://github.com/questdb/questdb/assets/37772591/55a7e4c0-d942-4c0f-b34d-82cfb1b32ea8)

hilmialf commented on core/src/main/java/io/questdb/PropServerConfiguration.java:
```diff
@@ -164,6 +164,7 @@ public class PropServerConfiguration implements ServerConfiguration {
     private final int maxUncommittedRows;
     private final MetricsConfiguration metricsConfiguration = new PropMetricsConfiguration();
     private final boolean metricsEnabled;
+    private final MicrosecondClock microsecondClock;
```
@puzpuzpuz I did that. But it reformat some other places too for example this MicrosecondClock got repositioned upon save.

puzpuzpuz replied:
That's ok. Occasionally, we get patches that aren't properly auto-formatted.

puzpuzpuz commented:
> In addition, I found the behavior of extend() for `MemoryCARWImpl` and `MemoryCMARWImpl` are different internally when the new size is multiple of pageSize. In current implementation, I somehow modify them so that they behave similarly. You can find the detail on the `SpillableHeapMemory` `checkAndExtend()` part. Please let me know if this is not what you thought it should be.

Left a few comments, including the one on the extension behavior.

hilmialf commented:
@puzpuzpuz Thanks a lot for taking your time to give a thorough review. That is really helpful for me. Will get back to you soon.
Additionally, I notice that the memory layout for FastMap has changed. Should I incorporate those changes to this PR?

hilmialf commented on core/src/main/java/io/questdb/cairo/map/SpillableHeapMemory.java:
```diff
@@ -0,0 +1,402 @@
+        mem = initMem();
+        spillPath = spillFilePath;
+
+        // pre-check spillPath can be used
```
Well, since ff.touch(spillPath) can throw exception, I thought pretest during initialization might be helpful rather than let it unnoticeable until by the time its started spilling and fail. wdyt?

puzpuzpuz replied:
No, most methods in `FilesFacade` don't throw. Instead, the result has to be checked. But my point is that this check is expensive considering that in most cases we don't expect disk spill to happen. Also, even if this check is successful, there is a chance that later file operations will fail.

hilmialf commented on core/src/main/java/io/questdb/cairo/map/SpillableHeapMemory.java:
```diff
@@ -0,0 +1,402 @@
+            return;
+        }
+
+        // MemoryCMARWImpl does not have memory breached checking, so adding one here. In addition, default
```
Hi Andrei,
Sorry for late follow up.
Id like to clarify more detail about the extend behavior since Ive been overthinking about this. To start, this is my current understanding:
- `pageSize` should be power of two
- in the in-mem mode, it grows as a power of two.
- `extendSegmentSize` should also be power of two.
- when it grows, it grows as a multiple of extendSegmentSize (thus the size of mmapped heap is not necessarily power of two)

Then, 
- In FastMap, the `MAX_HEAP_SIZE` is defined as `(Integer.toUnsignedLong(-1) - 1) << 3`. However, it is not a power of 2 (which might also not an exact multiples of `extendSegmentSize`). How do you think the extend behavior should be when extending up to `MAX_HEAP_SIZE`?
Is it okay to allow memory to extend beyond the MAX_HEAP_SIZE? And if it is allowed, is special handling required so that only part of the allocated memory can be used instead of all (in this case, we only use the memory up to MAX_HEAP_SIZE rather than its actual allocated size)

hilmialf replied:
In addition, lets say the requested size is n * pageSize, what should be the allocated size?
In current implementation, CARW and CMARW has different implementation.
- CARW: requested n*pageSize -> given n*pageSize
- CMARW: requested n*pageSize -> given (n+1)*pageSize

puzpuzpuz replied:
> In addition, lets say the requested size is `n * pageSize`, what should be the allocated size?

In file-based mode, we should grow the file in pages (page size should be multiple of the OS page, see `Files.PAGE_SIZE`). In in-memory mode, we should keep existing logic with the power of 2. In both cases, the allocated size should be equal or larger than the requested size.

hilmialf replied:
Thanks for the clarification. So, lets say the `extendSegmentSize` is set to 4GB (default value you proposed earlier) and the current heap size is 28GB (7 pages). When map is required to grow to 29GB, for example, since it extends based on page size, it will never be allowed since the next target size is 32GB = 34,359,738,368, while the max allowed heap is `Integer.toUnsignedLong(-1) - 1) << 3 = 34,359,738,352`.
In that case, is it okay to just throw error, or should it be allowed to grow to the MAX_HEAP_SIZE?

puzpuzpuz replied:
> In that case, is it okay to just throw error, or should it be allowed to grow to the MAX_HEAP_SIZE?

I'd say that if the requested size (29GB) is under `MAX_HEAP_SIZE`, we should grow to `MAX_HEAP_SIZE`. After that, the next growth attempt should fail.

> So, lets say the `extendSegmentSize` is set to 4GB (default value you proposed earlier)

I recall that I mentioned 4GB as the disk spill threshold, not the `extendSegmentSize`. I'd make `extendSegmentSize` something in the 256-512MB range by default. The idea is to do less costly file growth operations, while not growing it too rapidly to avoid using the disk space too eagerly.

hilmialf replied:
> I recall that I mentioned 4GB as the disk spill threshold, not the extendSegmentSize. I'd make extendSegmentSize something in the 256-512MB range by default. The idea is to do less costly file growth operations, while not growing it too rapidly to avoid using the disk space too eagerly.

I see

> I'd say that if the requested size (29GB) is under MAX_HEAP_SIZE, we should grow to MAX_HEAP_SIZE. After that, the next growth attempt should fail.

In summary, does that mean the last extension may extend without respecting the `extendSegmentSize`?
Or in other word, should we extend up to MAX_HEAP_SIZE, or up to the size below MAX_HEAP_SIZE which is multiple of `extendSegmentSize`?

puzpuzpuz replied:
> In summary, does that mean the last extension may extend without respecting the `extendSegmentSize`?
> Or in other word, should we extend up to MAX_HEAP_SIZE, or up to the size below MAX_HEAP_SIZE which is multiple of `extendSegmentSize`?

Yes, that sounds logical to me. File extension should try to use as much heap space as possible.

hilmialf commented on core/src/main/java/io/questdb/cairo/map/FastMap.java:
```diff
@@ -538,23 +594,20 @@ public void putBin(BinarySequence value) {
-            assert appendAddress + Byte.BYTES <= kLimit;
-            Unsafe.getUnsafe().putByte(appendAddress, (byte) (value ? 1 : 0));
-            appendAddress += Byte.BYTES;
+            mem.putBool(appendOffset, value);
```
Sure, will do!

hilmialf commented on core/src/main/java/io/questdb/cairo/map/FastMap.java:
```diff
@@ -651,41 +693,29 @@ private class VarSizeKey extends BaseKey {
-            }
+            // this looks like having different contract
+            // with that in MemoryCARW
+            // Here it only accepts 32bit, while mem accepts 2^64
```
alright!

hilmialf commented on core/src/main/java/io/questdb/cairo/map/FastMap.java:
```diff
@@ -211,14 +270,36 @@ public FastMap(
+            extendSegmentSize = maxSize;
+        }
+        spillFilePath = spillRoot != null ?
+                Path.getThreadLocal(spillRoot).concat(UUID.randomUUID() + ".tmp") :
```
After searching the code, I couldnt be sure where does the QDB's `Uuid` is exactly used to generate some ID. In a lot of places ive seen java UUID is still used instead.
Could you point me to it please?

puzpuzpuz replied:
> In a lot of places ive seen java UUID is still used instead.

Those are tests and benchmarks. We don't use std lib UUID in production code.

> Could you point me to it please?

Sure. Generating a random `Uuid` should be as simple as:
```java
long lo = rnd.nextLong();
long hi = rnd.nextLong();
Uuid uuid = new Uuid(lo, hi);
```

Here, `rnd` is an `Rnd` instance. We store these instances in `SqlExecutionContext#getRandom()`.

You may also refer to `RndUuidFunctionFactory`, but that class additionally sets UUID v4 version.

hilmialf commented on core/src/test/java/io/questdb/test/cairo/map/SpillableHeapMemoryTest.java:
```diff
@@ -0,0 +1,459 @@
+
+public class SpillableHeapMemoryTest extends AbstractCairoTest {
+    private static FilesFacade ff = new FilesFacadeImpl();
+    private static Path spillPath = Path.getThreadLocal(root).concat("fastmap.tmp").$();
```
I believe it also applies similarly when passing it down to class instantiation as an argument?

puzpuzpuz replied:
If the class constructor doesn't store the path as a field, but uses it locally, that's fine. We have `@Transient` annotation to mark arguments that shouldn't outlive the constructor call.
